pkg/lib/messaging/limits.go

```diff
@@ -51,7 +51,10 @@ type Limits struct {
 }
 
 func (l *Limits) cancel(ctx context.Context, reservation *Reservation) {
-	l.UsageLimiter.Cancel(ctx, reservation.UsageReservation)
+	// Usage reservation can be nil if usage limit reached
+	if reservation.UsageReservation != nil {
+		l.UsageLimiter.Cancel(ctx, reservation.UsageReservation)
+	}
```

Comment:
Instead of fixing here, we should make Cancel() handle nil reservation.

In pkg/lib/ratelimit/limiter.go Cancel() handle nil Reservation too.